database/sql/src/main/scala/cromwell/database/slick/MetadataSlickDatabase.scala

```diff
@@ -358,4 +362,51 @@ class MetadataSlickDatabase(originalDatabaseConfig: Config)
       labelAndKeyLabelValues, labelOrKeyLabelValues, excludeLabelAndValues, excludeLabelOrValues, submissionTimestampOption, startTimestampOption, endTimestampOption, metadataArchiveStatus, includeSubworkflows)
     runTransaction(action)
   }
+
+  override def deleteNonLabelMetadataForWorkflow(rootWorkflowId: String)(implicit ec: ExecutionContext): Future[Int] = {
+
+    val delete: DBIO[Int] = checkDeletionPreconditions(rootWorkflowId) flatMap { checkResult: Checked[Unit] =>
+      checkResult match {
+        case Right(_) =>
+          dataAccess.metadataEntriesWithoutLabelsForRootWorkflowId(rootWorkflowId).delete
+        case Left(errors: NonEmptyList[String]) =>
+          DBIO.failed(new Exception(errors.toList.mkString(",")))
+      }
+    }
+
+    runTransaction(delete)
+  }
+
+  private def checkDeletionPreconditions(rootWorkflowId: String)(implicit ec: ExecutionContext): DBIO[Checked[Unit]] = {
+    dataAccess.workflowMetadataSummaryEntriesForWorkflowExecutionUuid(rootWorkflowId).result.headOption map {
+      case Some(workflow: WorkflowMetadataSummaryEntry) =>
+        (isTerminalCheck(workflow), isRootWorkflowCheck(workflow)) mapN {
+          case (_, _) => ()
+        }
+      case None =>
+        s"""[Carbonite metadata deletion] Failed with non-existent root workflow "$rootWorkflowId"""".invalidNelCheck
+    }
+  }
+
+  private def isTerminalCheck(summaryEntry: WorkflowMetadataSummaryEntry): Checked[Unit] = {
+    import cromwell.core.WorkflowState
+
+    summaryEntry.workflowStatus match {
+      case Some(status) =>
+        if (WorkflowState.WorkflowStateValues.filter(_.isTerminal).map(_.toString).contains(status))
+          ().validNelCheck
+        else
+          s"""[Carbonite metadata deletion] Failed with non-terminal summary status "$status" for workflow "${summaryEntry.workflowExecutionUuid}"""".invalidNelCheck
+      case None =>
+        s"""[Carbonite metadata deletion] Failed because summary status unexpectedly empty for workflow "${summaryEntry.workflowExecutionUuid}"""".invalidNelCheck
+    }
+  }
+
+  private def isRootWorkflowCheck(summaryEntry: WorkflowMetadataSummaryEntry): Checked[Unit] = {
+    if (summaryEntry.parentWorkflowExecutionUuid.isEmpty && summaryEntry.rootWorkflowExecutionUuid.isEmpty)
```

Comment:
Similarly, the SQL on the server, not the scala running in the JVM, should only delete rows where the parent and the root are null.

Reply:
I think this comment may not account for the fact that the metadata entries themselves do not contain root/parent workflow IDs in a usable way (i.e. they are metadata values with no index)?